app/controllers/adventures_controller.rb

```diff
@@ -103,7 +117,16 @@ def adventure_params
       :slug,
       :description,
       :public,
-      :theme
+      :theme,
+      :password,
+      :password_confirmation,
+      :has_password
     )
   end
+
+  def check_authentication
+    unless session[@adventure.title] || !@adventure.has_password
```

Comment:
i'll leave this decision to you, but i tend to avoid the `unless` conditional _unless_ I can stick it at the end of a line (similar to how this sentence is structured). Mainly because thinking about negative cases hurt my brain. So this I'd probably go with:
```rb
if @adventure.has_password? && !session[@adventure.title]
```